hal/spi_api.h

```diff
@@ -19,220 +18,346 @@
 #ifndef MBED_SPI_API_H
 #define MBED_SPI_API_H
 
+#include <stdbool.h>
+#include <stdint.h>
 #include "device.h"
-#include "hal/dma_api.h"
-#include "hal/buffer.h"
 
 #if DEVICE_SPI
+/**
+ * \defgroup hal_spi SPI: Serial peripheral interface HAL API.
+ * Low level interface to the serial peripheral interface of a target.
+ *
+ * # Programming model
+ * This is an OOP like API. spi_t is used a "this".<br/>
+ * Methods not having this type as their first parameter are **class** methods.<br/>
+ * Methods with an spi_t object as their first parameter are **instance** methods.<br/>
+ * `spi_init()` acts as the "constructor"(/initialiser) of this *class*.<br/>
+ * `spi_free()` acts as the "destructor"(/deinitialiser) of this *class*.<br/>
+ *
+ * # Defined behaviours
+ * - Multiple instances can share pins such as MISO/MOSI/MCLK but not their SS pin. This pin is used
+ *   to distinguish communication channels (master or slaves).
+ * - As the instance allocation is at the user's responsibility, the hal implementation cannot keep
```

Comment:
the other way to address that is to tell the user that he must **NOT** move and/or deallocate the spi_t or any other object passed to the hal_impl as it may keep a pointer to them.

```
 * - As the hal implementation may need to keep an alias to any value in a spi_t instance, the user
 *   must **NOT** move and/or de-allocated the structure before calling spi_free() on this instance.
```
Should this be part of the defined or undefined behaviour ? Maybe the programming model ?